src/Raven.Client/Http/RequestExecuter.cs

```diff
@@ -296,41 +317,91 @@ private void UpdateTopologyCallback(object _)
-
             if (command.FailedNodes == null)
                 command.FailedNodes = new HashSet<ServerNode>();
 
+            node.IsFailed = true;
             command.FailedNodes.Add(node);
 
-            var failoverNode = GetFailoverNode(command);
-
-            if (failoverNode != null)
-            {
-                await ExecuteAsync(failoverNode, context, command);
-                return;
-            }
-
-            throw new HttpRequestException("Tried all nodes in the cluster but failed getting a response", e);
+            var failoverNode = ChooseNodeForRequset(command, e);
+            await ExecuteAsync(failoverNode, context, command);
         }
 
-        private ServerNode GetFailoverNode<T>(RavenCommand<T> command)
+        private ServerNode ChooseNodeForRequset<T>(RavenCommand<T> command, HttpRequestException exception = null)
         {
-            //TODO: implement failover policies
             var topology = _topology;
 
-            if (command.FailedNodes == null)
-                command.FailedNodes = new HashSet<ServerNode>();
-
             var leaderNode = topology.LeaderNode;
-            if (command.FailedNodes.Contains(leaderNode) == false)
-                return leaderNode;
 
-            foreach (var node in topology.Nodes)
+            if (command.IsReadRequest)
+            {
+                if (topology.ReadBehavior == ReadBehavior.Leader)
+                {
+                    if (command.IsFailedWithNode(leaderNode) == false)
+                        return leaderNode;
+                    throw new HttpRequestException("Leader not was failed to make this request. The current ReadBehavior is set to Leader to we won't failover to a differnt node.", exception);
+                }
+
+                if (topology.ReadBehavior == ReadBehavior.All)
+                {
+                    if (leaderNode.IsFailed == false && command.IsFailedWithNode(leaderNode) == false)
+                        return leaderNode;
+
+                    // TODO: Should we choose nodes here by rate value as for SLA?
+                    foreach (var node in topology.Nodes)
+                    {
+                        if (node.IsFailed == false && command.IsFailedWithNode(leaderNode) == false)
+                            return node;
+                    }
+
+                    throw new HttpRequestException("Tried all nodes in the cluster but failed getting a response", exception);
+                }
+
+                if (topology.ReadBehavior == ReadBehavior.SLA)
+                {
+                    if (leaderNode.IsFailed == false && command.IsFailedWithNode(leaderNode) == false && leaderNode.IsRateSurpassed(topology.SLA.RequestTimeThresholdInMilliseconds))
+                        return leaderNode;
+
+                    var nodesWithLeader = topology.Nodes;
+                    nodesWithLeader.Add(leaderNode);
+
+                    leaderNode.DecreaseRate(topology.SLA.RequestTimeThresholdInMilliseconds);
```

Comment:
This will not work as you would expect. The ComplexTimeMetric was increasing the metric for the current node that we send the request, but decreasing for the others that we tried before that (check the AddCurrent method there). By doing this we are slowly decreasing the EWMA rate so after some time we can get back to other nodes. With your approach when all other nodes will surpass the threshold then you will never get back go them because you will never decrease their EWMA rate.